Feature/v2 updates

fivetran-joemarkiewicz opened the pull request:
**Are you a current Fivetran customer?** 
<!--- Please tell us your name, title and company -->
Fivetran created PR

**What change(s) does this PR introduce?** 
<!--- Describe what changes your PR introduces to the package and how to leverage this new feature. -->
This PR applies the Ad Reporting V2 updates. In particular this PR does the following:
- Adds the `ad_group_criterion_history` source and accompanying staging models
- Adds the `ad_group_stats` source and accompanying staging models
- Adds the `campaign_stats` source and accompanying staging models
- Adds the `account_stats` source and accompanying staging models
- Adds the `keyword_stats` source and accompanying staging models
- Adds tests to each of the new and existing end models
- Completely removes the Adwords versions of the package as it has been fully sunset
- Applies README updates
- Adds identifiers to all source models

**Did you update the CHANGELOG?** 
<!--- Please update the new package version’s CHANGELOG entry detailing the changes included in this PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [ ] Yes

**Does this PR introduce a breaking change?**
<!--- Does this PR introduce changes that will cause current package users' jobs to fail or require a `--full-refresh`? -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [X] Yes (please provide breaking change details below.)
- [ ] No  (please provide an explanation as to how the change is non-breaking below.)

This will be a major update and will be breaking for existing users.

**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 
<!--- The dbt_project.yml and the integration_tests/dbt_project.yml files contain the version number. Be sure to upgrade it accordingly -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [X] Yes

**Is this PR in response to a previously created Bug or Feature Request**
<!--- If an Issue was created it is helpful to track the progress by linking it in the PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [ ] Yes, Issue/Feature [link bug/feature number here]
- [X] No 

**How did you test the PR changes?** 
<!--- Proof of testing is required in order for the PR to be approved. -->
<!--- To check a box, remove the space and insert an x in the box (eg. [x] CircleCi). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [X] CircleCi <!--- CircleCi testing is only applicable to Fivetran employees. --> 
- [X] Local (please provide additional testing details below)

**Select which warehouse(s) were used to test the PR**
<!--- To check a warehouse remove the space and insert an x in the box (eg. [x] Bigquery). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [X] BigQuery
- [X] Redshift
- [X] Snowflake
- [X] Postgres
- [X] Databricks
- [ ] Other (provide details below)

**Provide an emoji that best describes your current mood**
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
✌️ 

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

fivetran-joemarkiewicz committed:
feature/v2_updates

fivetran-joemarkiewicz committed:
campaign trial type removal

fivetran-joemarkiewicz committed:
keyword and search adjustments

fivetran-joemarkiewicz committed:
validation updates

fivetran-joemarkiewicz committed:
seed data adjustments

fivetran-joemarkiewicz committed:
additional seed fixes

fivetran-joemarkiewicz committed:
seed schema changed

fivetran-joemarkiewicz committed:
readme and other docs updates

fivetran-joemarkiewicz committed:
moving auto tagging variable to transform package

fivetran-joemarkiewicz committed:
changelog update

fivetran-joemarkiewicz committed:
minor pre-release updates

fivetran-joemarkiewicz committed:
auto tagging variable tests moved to transform

fivetran-joemarkiewicz commented:
FYI the base branch is `feature/google-ads-api-default` as that branch will be merged prior to this one and I want to make sure I am not duplicating reviews from the changes made in PR #28

fivetran-joemarkiewicz committed:
macro field removals

fivetran-joemarkiewicz committed:
macro fix

fivetran-joemarkiewicz committed:
adgroup adjustment for ad and utm

fivetran-joemarkiewicz committed:
add coalesce

fivetran-joemarkiewicz committed:
spelling fix

fivetran-joemarkiewicz committed:
adgroupid add

fivetran-joemarkiewicz committed:
decision log and minor adjustments

fivetran-sheringuyen commented on models/stg_google_ads__account_history.sql:
```diff
@@ -24,9 +23,11 @@ final as (
     select 
         id as account_id,
-        updated_at,
+        updated_at as updated_timestamp,
```
I'm interested in your preference for `updated_at` vs `updated_timestamp` and whether you think this is worth noting in our style guide. 🤔

fivetran-joemarkiewicz replied:
It looks like this is the only one of our packages that does this `_timestamp` name change. I just switched back to just use the `_at` convention to be consistent with our other packages. Additionally, I am not entirely sold on the requirement of this naming change.

fivetran-joemarkiewicz replied:
Made the change to the models and .yml documentation.

fivetran-sheringuyen commented on models/stg_google_ads__account_stats.sql:
```diff
@@ -0,0 +1,38 @@
+        cost_micros / 1000000.0 as spend, 
+        impressions
+        
+        {% for metric in var('google_ads__account_stats_passthrough_metrics', []) %}
```
I faintly remember a discussion the team had about the usage of `var('google_ads__account_stats_passthrough_metrics', [])` vs more explicit like the use of lengths or something similar, but I can't remember which package that new logic was in. I think it was actually a package that you created a PR for but I reviewed it..does this ring a bell? (I should have definitely noted this better somewhere 😅 )

fivetran-joemarkiewicz replied:
Good call out! That conversation was regarding our use of adding passthrough columns with an `if` statement. 
https://github.com/fivetran/dbt_zendesk_source/blob/aef4f3e6d436b4d5f52ecac81c8beb71951289b0/models/stg_zendesk__ticket.sql#L25-L30

I actually did not realize we also use the `for loop` version until I came across our use of them within these ad packages. Personally, I like the `for loop` better as it reads a bit more naturally and is explicit by default (imo). The `if` statement is pretty clunky and hard to follow, whereas the `for loop` naturally tells me to iterate through the variable without needing a condition and just sets a default to be safe.

I would like to actually suggest we start using the for loop method from now on.

fivetran-sheringuyen replied:
I agree, I also like the for loop method better! Will add this to the dev philo. Thanks Joe!

fivetran-sheringuyen commented on CHANGELOG.md:
```diff
@@ -1,3 +1,29 @@
+
+- README updates for easier navigation and use of the package. ([#29](https://github.com/fivetran/dbt_google_ads_source/pull/29))
+- Addition of identifier variables for each of the source tables to allow for further flexibility in source table direction within the dbt project. ([#29](https://github.com/fivetran/dbt_google_ads_source/pull/29))
+- Included grain uniqueness tests for each staging table. ([#29](https://github.com/fivetran/dbt_google_ads_source/pull/29))
```
Is there a preference of referring to the same PR multiple times within the `CHANGELOG` vs outlining just once at the top which PR all the changes were made in?

fivetran-joemarkiewicz replied:
Yeah I saw that you highlighted all of the changes were included in the single PR for your Changelog. I think that is a good call. When there are multiple PRs or just one minor update that is tied to a single PR then we can take the approach I have done here. 

I will update my CHANGELOG to match that style.

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
+  - Name columns for consistency across all packages and for easier analysis
+  - Adds freshness tests to source data
+  - Adds column-level testing where applicable. For example, all primary keys are tested for uniqueness and non-null values.
+- Generates a comprehensive data dictionary of your google_ads data through the [dbt docs site](https://fivetran.github.io/dbt_google_ads_source/).
```
Maybe "Google Ads"?

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
+# 📣 What does this dbt package do?
+- Materializes [Google Ads staging tables](https://fivetran.github.io/dbt_google_ads_source/#!/overview/google_ads_source/models/?g_v=1&g_e=seeds) which leverage data in the format described by [this ERD](https://fivetran.com/docs/applications/google-ads#schemainformation). These staging tables clean, test, and prepare your google_ads data from [Fivetran's connector](https://fivetran.com/docs/app ...
+  - Name columns for consistency across all packages and for easier analysis
+  - Adds freshness tests to source data
```
I forgot to add this too -- but it looks like there is currently no freshness test for the Google source data. 

For the Apple Store & Google Play package, we used the below:
>     freshness: 
>       warn_after: {count: 48, period: hour}
>       error_after: {count: 168, period: hour}

Would you say this is also a good freshness standard for our Ad Reporting as well or would you suggest something different?

fivetran-joemarkiewicz replied:
Yeah I think that is a good standard! Let's use that going forward and I will add the tests to this package.

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen commented on models/src_google_ads.yml:
```diff
@@ -1,16 +1,17 @@
 version: 2
 
```
Add freshness tests

fivetran-joemarkiewicz replied:
Added

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
-    google_ads__url_passthrough_metrics: ['the', 'list', 'of', 'metric', 'columns', 'to', 'include'] # from adwords.final_url_performance
-    google_ads__criteria_passthrough_metrics: ['the', 'list', 'of', 'metric', 'columns', 'to', 'include'] # from adwords.criteria_performance
-
     # If you're using the Google Ads API source
```
We could probably remove this comment now!

fivetran-joemarkiewicz replied:
Removed

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
-### Changing the Build Schema
-By default this package will build the Google Ads staging models within a schema titled (<target_schema> + `_stg_google_ads`) in your target database. If this is not where you would like your Google Ads  staging data to be written to, add the following configuration to your `dbt_project.yml` file:
+### Change the build schema
+By default, this package builds the google_ads staging models within a schema titled (`<target_schema>` + `_stg_google_ads`) in your destination. If this is not where you would like your google_ads staging data to be written to, add the following configuration to your root `dbt_project.yml` file:
```
Maybe instead of "google_ads" we use "Google Ads"?

> By default, this package builds the Google Ads staging models within a schema titled (`<target_schema>` + `_stg_google_ads`) in your destination. If this is not where you would like your Google Ads staging data to be written to, add the following configuration to your root `dbt_project.yml` file:

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
-
-The package assumes that the corresponding destination tables are named `final_url_performance`, `criteria_performance`, and `click_performance` respectively. If these tables have different names in your destination, enter the correct table names in the `google_ads__final_url_performance`, `google_ads__click_performance`, and `google_ads__criteria_performance` variables so that the package can fi ...
+## Step 3: Define database and schema variables
+By default, this package runs using your destination and the `google_ads` schema. If this is not where your google_ads data is (for example, if your google_ads schema is named `google_ads_fivetran`), add the following configuration to your root `dbt_project.yml` file:
```
> If this is not where your Google Ads data is ...

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen reviewed (commented):
Hey @fivetran-joemarkiewicz, great job on the deprecation and rest of the PR! I just had some notes I wanted us to keep track of and a couple in-line comments.

There may be some merge conflicts since the `feature/google-ads-api-default` updates, but I want to make sure certain things don't accidentally get lost when we do get around to releasing this package:
- [x] Documentation for final urls regex matching - specifically in the `stg_google_ads__ad_history.source_final_urls` tests.
- [x] Removing deprecation notes from `README`

It looks like in both `google_ads` and `google_ads_source`, `run_results.json` is missing. Is this intentional (and should be updated in the `generate_docs` macro)?

fivetran-sheringuyen commented on dbt_project.yml:
```diff
@@ -1,29 +1,29 @@
 
 models:
   google_ads_source:
     +schema: stg_google_ads
```
+schema: google_ads_source (to reduce redundancy)

fivetran-joemarkiewicz replied:
Updated

fivetran-sheringuyen commented on README.md:
```diff
@@ -1,192 +1,115 @@
-### Changing the Build Schema
-By default this package will build the Google Ads staging models within a schema titled (<target_schema> + `_stg_google_ads`) in your target database. If this is not where you would like your Google Ads  staging data to be written to, add the following configuration to your `dbt_project.yml` file:
+### Change the build schema
+By default, this package builds the google_ads staging models within a schema titled (`<target_schema>` + `_stg_google_ads`) in your destination. If this is not where you would like your google_ads staging data to be written to, add the following configuration to your root `dbt_project.yml` file:
```
With changing the redundancy:

> <target_schema> + `_google_ads_source`

fivetran-joemarkiewicz replied:
Updated

fivetran-joemarkiewicz committed:
review notes updates

fivetran-joemarkiewicz committed:
added yml documentation for test warning

fivetran-joemarkiewicz committed:
merge conflicts resolution

fivetran-joemarkiewicz committed:
minor README edit

fivetran-joemarkiewicz commented:
Thanks for the review @fivetran-sheringuyen! I addressed your inline comments and only had a question on the one I left unresolved regarding the passthrough logic. Let me know your thoughts! Additionally I addressed the following that was not captured inline:
- Added the yml documentation for the test warning with the `stg_google_ads__ad_hsitory` final_urls field.
- Included the `run_results.json`. I initially intentionally omitted it as it isn't necessarily needed. However, I noticed the datatype value is missing within the docs if you omit this file. Therefore I added it back in.
- Addressed merge conflicts after switching the base branch to be `main`.
- Regenerated the docs.

fivetran-sheringuyen reviewed (approved):
Thanks @fivetran-joemarkiewicz for making the updates, this package looks good to me for release when we roll out Ad Reporting V2!

fivetran-joemarkiewicz commented:
Things to update before merging:
- [x] Add not null tests
- [x] Ensure passthrough metrics is correct in both CTEs and no duplicates in macro

fivetran-joemarkiewicz committed:
Update packages.yml

fivetran-joemarkiewicz committed:
final before merge updates

fivetran-joemarkiewicz committed:
source config code reuction and docs generation

fivetran-joemarkiewicz committed:
added name to ad history

fivetran-sheringuyen committed:
removing fivetran_synced from macros

fivetran-sheringuyen committed:
seed data

fivetran-sheringuyen committed:
macro update

fivetran-sheringuyen committed:
seed data

fivetran-sheringuyen committed:
seed data

fivetran-sheringuyen committed:
updating spark split

fivetran-sheringuyen committed:
spark split function

fivetran-sheringuyen committed:
spark split

fivetran-sheringuyen committed:
readme
